### ISSUE TITLE
Reference types crash wasm-opt‘s asyncify pass

### ISSUE THREAD

**Issue by surma** [WebAssembly/binaryen#3739]
```wat
(module
  (type (;0;) (func (param i32 i32)))
  (type (;1;) (func (result externref)))
  (import "gl" "createProgram" (func $src/asc/gl/createProgram (type 1)))
  (func $src/asc/lemonsoda/init (type 0) (param i32 i32)
    call $src/asc/gl/createProgram
    drop
  )
  (export "init" (func $src/asc/lemonsoda/init))
)
```

```
$ wat2wasm --enable-reference-types test.wat -o test.wasm
$ wasm-opt --asyncify --enable-reference-types test.wasm -o test2.wasm
invalid type
UNREACHABLE executed at /tmp/binaryen-20210320-71286-1g5tfwr/binaryen-version_100/src/wasm/wasm-type.cpp:488!
[1]    56081 abort      wasm-opt --asyncify --enable-reference-types test.wasm -o test2.wasm
```

**Comment by surma** [comment]
It seems to be specifically about using `externref` in imported functions.

**Comment by martianboy** [comment]
FWIW, this is not limited to `externref`, it happens for all reference types since they can't be stored on memory. 
Just out of curiosity, @kripken would this entail setting the reftype value on a table and store the index instead?

**Comment by surma** [comment]
@martianboy Ah of course! That makes a lot of sense.

Since the reference types proposal also allows having multiple tables, it seems like your idea with the table is a really good one.

**Comment by martianboy** [comment]
@surma Right, although this cannot happen yet in Binaryen, as it still doesn't have table operations and typed tables. But it shouldn't be too difficult to support reference types in Asyncify after those are done. I may still be missing something though.

**Comment by kripken** [comment]
I think that's correct @martianboy & @surma. Yes, the issue is writing references to linear memory. Instead we would need to create a new table in asyncify, and use that to spill values using table.get/table.set, etc., so as mentioned above we'd need support for those features.

**Comment by martianboy** [comment]
Thanks for confirmation, @kripken. I'm in the middle of adding types to tables and element segments and after that I was planning to add table operations. I can see if I can fix Asyncify next. :)

**Comment by kripken** [comment]
Sounds great @martianboy ! I hope that's pretty straightforward when you get to it (only nontrivial part might be to have two "stack pointers"), let me know if you have questions I can help with.

**Comment by surma** [comment]
Not to be _that_ impatient dude on GH, but curious if there are any updates on this?

**Comment by martianboy** [comment]
@surma So I'm "almost" done with implementation of table instructions which was a prerequisite. My guess is that I will be able to get to Asyncify in about 10 days. Sorry for the wait. :)

**Comment by surma** [comment]
Oh please, don’t apologize! I’m just curious! Very excited to see this :)

**Comment by alexdoesh** [comment]
@kripken looks like the problem is still exists, can we elaborate a little bit more on how it should be solved, I'd be happy to spend time for this. any real blockers, or just a question of implementation?

**Comment by kripken** [comment]
@alexdoesh I think we know how it should work, looks like a summary is in

https://github.com/WebAssembly/binaryen/issues/3739#issuecomment-808330306

In more detail, we spill stuff to a buffer in linear memory normally, in the existing code. For reference types we need a buffer in a table that we can write references to. So we'd need to add such a table to the wasm, and add some code that manages allocating buffers in it, and then add code to allocate the buffer and read/write from it (that is, if a local is a reference, we'd write to the table instead of linear memory).

If you want to do it that would be great!

**Comment by kripken** [comment]
@alexdoesh 

Yes, that looks like the right idea!

Aside from polish and the other stuff you already mentioned, the only other thing is I wonder if we want to support more than one pause at a time. We do that currently, and in Emscripten the Fibers API uses it. To support that here, we couldn't use absolute offsets in the table, but would need to allocate regions in it, which would add significant complexity. Also I think atm the user allocates the main buffer in linear memory, but here we would need to do it in Asyncify (since the user isn't even aware of the new table added here). It's probably best to not do that for now, but we should document the limitation, and perhaps there's a way we can assert at runtime.

**Comment by alexdoesh** [comment]
@kripken I had the feeling I'm not handling it properly, makes more sense now, thank you.

As I understand, we already have a helper to diagnose how many local variables (potentially affected by stack rewind) we have, I wanted to use it (or similar approach) anyway, it would allow us to understand the size of the region + what tables we would need (per ref-type) + some logic to "isolate" region per function. Does it seem enough? Or I misunderstood the meaning of "regions" in the ctx?

Do you think it's ok for asyncify pass to create a table if this would be explicit to the end user? Using a command line argument, for example (`--reference-types-support` or smth like this, and/or ability to provide a tables name prefix).

I'm not a C++ dev (I'm writing python compiler to wasm in python, inspired by asc), but still want to tackle the problem (if somebody would be patient to assist/review ofc), it looks pretty important at least for my pet project.

**Comment by kripken** [comment]
@alexdoesh Yes, I think the current approach is enough for now. Just a comment mentioning the multi-pause issue sounds good.

Yes, I think it's the right thing to create a table here. We can do that automatically/silently in this pass. That should not be observable to the user (at least not unless they read the wasm :smile: ) since the table is not exported or imported.

I think with some cleanup that code can land. Good work!

**Comment by K0IN** [comment]
Any updates on this?
@alexdoesh i tried your fork but it doesnt seem to work? 
Am I missing something here?

**Comment by izissise** [comment]
Hello, I've ended up here, is there any plan to merge the above pr?

**Comment by kripken** [comment]
@izissise Do you have a use case that needs it? It mainly makes sense for places with wasm exceptions but without JSPI (as JSPI is better than Asyncify), and the number of such places is declining according to

https://webassembly.org/features/

Chrome launched both features in the same release. Safari has both enabled in the last few preview builds. Firefox has JSPI behind a flag atm but hopefully will ship soon.

If the PR were simple we would just merge it even for a short-term benefit, but unfortunately adding that support includes some awkward things ("relaxed Flatten mode") that add complexity.

**Comment by izissise** [comment]
Thanks for the quick response, I'm trying to run turso database (sqlite) in a service worker.
I didn't know about jspi, I will try with it especially since it looks simpler with broswer support

**Comment by sbc100** [comment]
I'm running into this issue trying use externref in emscripten for the first time.   I'd love to see this fixed.  Is it close?

**Comment by sbc100** [comment]
Basically I'm trying to write code that works for both ASYNCIFY=1 and ASYNCIFY=2 (JSPI) in emscripten.     If we don't fix this issue then I would need to have a completely different codepath for ASYNCIFY=1 that doesn't use reference types at all.

**Comment by sbc100** [comment]
See https://github.com/emscripten-core/emscripten/pull/26964
